nservicebus/sagas/concurrency.md

```diff
@@ -1,48 +1,58 @@
+## Starting a saga
 
-## Concurrent access to non-existing saga instances
+When messages that start the same logical saga are handled simultaneously, NServiceBus ensures that only one message starts the saga. This is done in the same transaction as receiving the message and sending new messages.
 
-Sagas are started by the message types that a handled with `IAmStartedByMessages<T>`. If messages mapped to the same saga instance are processed concurrently there is a risk that duplicates of the instance will be created.
-
-In this case only one message is allowed to complete processing. The others roll back and the built-in retries in NServiceBus kick in. On the next retry, the saga instance is found, the race condition is solved, and that saga instance is updated instead, see below.
+Only one message handler is allowed to succeed. The saga state is created and sent messages are dispatched. The other message handlers fail, roll back, and their messages enter [recoverability](/nservicebus/recoverability/). When those messages are retried, the existing saga state is found. Handling the messages may involve changing the state or completing the saga. See below for how those scenarios are handled.
 
 partial: unique
 
-## Concurrent access to existing saga instances
+## Changes to saga state
+
+When messages are handled simultaneously by the same logical saga, and involve changing the saga state, NServiceBus ensures that only one message handler changes the state in the same transaction as receiving the message and sending any new messages.
+
+When using some persisters, the messages are handled in serial, one after another. In this case, all message handlers are allowed to succeed.
+
+When using other persisters, only one message handler is allowed to succeed. The saga state is updated, and sent messages are dispatched. The other message handlers fail, roll back, and their messages enter [recoverability](/nservicebus/recoverability/). When those messages are retried, the process repeats. One or more of those messages may attempt to complete the saga. See below for how that scenario is handled.
+
+See the [documentation for each persister](/persistence/) for details of which method is used.
+
+## Completing a saga
+
+When messages are handled simultaneously by the same logical saga, and the handling of at least one message involves completing the saga, NServiceBus ensures that only one message handler completes the saga. This is done in the same transaction as receiving the message and sending any new messages.
 
-When messages concurrently tries to update the same saga instance the storage will either detect and throw a concurrency exception or serialize access to the instance. Concurrency exceptions will be automatically resolved by the NServiceBus retries will.
+When using some persisters, the messages are handled in serial, one after another.
```

Comment:
Is "in serial" common usage? It sounds awkward to me.

Reply:
Also, its not correct, only saga state is persisted sequentially. Saga handlers can be processed in parallel. 
```suggestion
When using a persister that support pessimistic locking, saga handlers invocations on the same saga instance (not type) are sequential, one after another due.
```

Reply:
> Is "in serial" common usage? It sounds awkward to me.

Probably not. I'll change it.

Reply:
> Also, its not correct, only saga state is persisted sequentially. Saga handlers can be processed in parallel.

I'm not sure that's correct, and your suggestion doesn't suggest that either. Because pessimistic locking is used, the handler are effectively executed one after another, since the method is only invoked after the data has been read, which doesn't occur until after another handler has finished and the data from that handler is persisted. If that were not the case, and they were executed in parallel, then they would have dirty reads, and the pessimistic locking would not be occurring.